controller/engine_image_controller.go

```diff
@@ -378,9 +384,193 @@ func (ic *EngineImageController) syncEngineImage(key string) (err error) {
+//   1. Volume is not in engine upgrading process
+//   2. New default engine image version is newer than the current volume's engine version
+//   3. Volume is in attached state and it is able to do live upgrade
+func (ic *EngineImageController) getVolumesForEngineImageUpgrading(volumes map[string]*longhorn.Volume, newEngineImageResource *longhorn.EngineImage) ([]*longhorn.Volume, error) {
+	candiates := []*longhorn.Volume{}
+
+	for _, v := range volumes {
+		// Already attempted to upgrade volume's engine image
+		if v.Spec.EngineImage == newEngineImageResource.Spec.Image {
+			continue
+		}
+		// volume is in upgrading process to a different engine image. Need to wait for the upgrade to finish.
+		if v.Spec.EngineImage != v.Status.CurrentImage && newEngineImageResource.Spec.Image != v.Status.CurrentImage {
+			ic.logger.Infof("getVolumesForEngineImageUpgrading: upgrading in process for volume %v engine image from %v to %v, cannot upgrade to another engine image",
+				v.Name, v.Status.CurrentImage, v.Spec.EngineImage)
+			continue
+		}
+		// compare engine image version to make sure Longhorn doesn't automatically downgrade engine image
+		if !checkEngineImageSemanticVersion(v.Status.CurrentImage, newEngineImageResource.Spec.Image) {
+			continue
+		}
+		if !ic.canDoOfflineEngineImageUpgrade(v, newEngineImageResource) && !ic.canDoLiveEngineImageUpgrade(v, newEngineImageResource) {
+			continue
+		}
+		candiates = append(candiates, v)
+	}
+
+	return candiates, nil
+}
+
+// checkEngineImageSemanticVersion check if oldEngineImage and newEngineImage are valid semantic versions
+// and oldEngineImage is less than newEngineImage
+func checkEngineImageSemanticVersion(oldEngineImage, newEngineImage string) bool {
+	oldTag := getImageTag(oldEngineImage)
+	newTag := getImageTag(newEngineImage)
+	if oldTag == "master" {
+		oldTag = maxEngineImageVersion
+	}
+	if newTag == "master" {
+		newTag = maxEngineImageVersion
+	}
+	oldVersion, err := semver.NewVersion(oldTag)
+	if err != nil {
+		return false
+	}
+	newVersion, err := semver.NewVersion(newTag)
+	if err != nil {
+		return false
+	}
+	return oldVersion.LessThan(newVersion)
+}
+
+func getImageTag(image string) string {
+	tokens := strings.Split(image, ":")
+	if len(tokens) != 2 {
+		return ""
+	}
+	return tokens[1]
+}
+
+func (ic *EngineImageController) canDoOfflineEngineImageUpgrade(v *longhorn.Volume, newEngineImageResource *longhorn.EngineImage) bool {
+	return v.Status.State == types.VolumeStateDetached && newEngineImageResource.Status.State == types.EngineImageStateReady
+}
+
+// canDoLiveEngineImageUpgrade check if it is possible to do live engine upgrade for a volume
+// A volume can do live engine upgrade when:
+//   1. Volume is attached AND
+//   2. Volume's robustness is either healthy or degraded AND
+//   3. Volume's robustness is healthy if volume is DR volume
+//   4. Volume is not expanding AND
+//   5. The current volume's IM image version is compatible with the new engine image
```

Comment:
You are comparing the old engine image version with that of the new engine image here

Reply:
@shuo-wu 
> What if a user is using the customized/modified engine image? In this case, the version can be the same but the repository may be different. Or the image tag doesn't match the official tag partten.

We only consider the image's tag when comparing images. If the tags don't have valid semantic version format, we do nothing and thus this is a fail safe. 


> we can simply consider the default engine image as the latest engine image

1. We allow user do deploy additional images in Longhorn UI. User might deploy newer images. In this case, we cannot constantly force the newer images back to the default image. 
2. I am thinking about a failed-to-uprade case where user rollback Longhorn. In this case, the default image is not the newest one
3. This feature is design for end user who are not supposed to change Longhorn version (image tags). I think most users are doing that. 
4. One side note, Rancher new UI automatically add 2 more digit into the tag, this doesn't change the correctness of this design.

Reply:
Thank you. Fixed